// jest-snapshot/build/index.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
Object.defineProperty(exports, 'EXTENSION', {
  enumerable: true,
  get: function () {
    return _SnapshotResolver.EXTENSION;
  }
});
Object.defineProperty(exports, 'SnapshotState', {
  enumerable: true,
  get: function () {
    return _State.default;
  }
});
Object.defineProperty(exports, 'addSerializer', {
  enumerable: true,
  get: function () {
    return _plugins.addSerializer;
  }
});
Object.defineProperty(exports, 'buildSnapshotResolver', {
  enumerable: true,
  get: function () {
    return _SnapshotResolver.buildSnapshotResolver;
  }
});
exports.cleanup = void 0;
Object.defineProperty(exports, 'getSerializers', {
  enumerable: true,
  get: function () {
    return _plugins.getSerializers;
  }
});
Object.defineProperty(exports, 'isSnapshotPath', {
  enumerable: true,
  get: function () {
    return _SnapshotResolver.isSnapshotPath;
  }
});
exports.toThrowErrorMatchingSnapshot =
  exports.toThrowErrorMatchingInlineSnapshot =
  exports.toMatchSnapshot =
  exports.toMatchInlineSnapshot =
    void 0;
var fs = _interopRequireWildcard(require('graceful-fs'));
var _jestMatcherUtils = require('jest-matcher-utils');
var _SnapshotResolver = require('./SnapshotResolver');
var _printSnapshot = require('./printSnapshot');
var _utils = require('./utils');
var _plugins = require('./plugins');
var _State = _interopRequireDefault(require('./State'));
function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {default: obj};
}
function _getRequireWildcardCache(nodeInterop) {
  if (typeof WeakMap !== 'function') return null;
  var cacheBabelInterop = new WeakMap();
  var cacheNodeInterop = new WeakMap();
  return (_getRequireWildcardCache = function (nodeInterop) {
    return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
  })(nodeInterop);
}
function _interopRequireWildcard(obj, nodeInterop) {
  if (!nodeInterop && obj && obj.__esModule) {
    return obj;
  }
  if (obj === null || (typeof obj !== 'object' && typeof obj !== 'function')) {
    return {default: obj};
  }
  var cache = _getRequireWildcardCache(nodeInterop);
  if (cache && cache.has(obj)) {
    return cache.get(obj);
  }
  var newObj = {};
  var hasPropertyDescriptor =
    Object.defineProperty && Object.getOwnPropertyDescriptor;
  for (var key in obj) {
    if (key !== 'default' && Object.prototype.hasOwnProperty.call(obj, key)) {
      var desc = hasPropertyDescriptor
        ? Object.getOwnPropertyDescriptor(obj, key)
        : null;
      if (desc && (desc.get || desc.set)) {
        Object.defineProperty(newObj, key, desc);
      } else {
        newObj[key] = obj[key];
      }
    }
  }
  newObj.default = obj;
  if (cache) {
    cache.set(obj, newObj);
  }
  return newObj;
}
var Symbol = globalThis['jest-symbol-do-not-touch'] || globalThis.Symbol;
var Symbol = globalThis['jest-symbol-do-not-touch'] || globalThis.Symbol;
var jestExistsFile =
  globalThis[Symbol.for('jest-native-exists-file')] || fs.existsSync;
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
const DID_NOT_THROW = 'Received function did not throw'; // same as toThrow
const NOT_SNAPSHOT_MATCHERS = `Snapshot matchers cannot be used with ${(0,
_jestMatcherUtils.BOLD_WEIGHT)('not')}`;
const INDENTATION_REGEX = /^([^\S\n]*)\S/m;

// Display name in report when matcher fails same as in snapshot file,
// but with optional hint argument in bold weight.
const printSnapshotName = (concatenatedBlockNames = '', hint = '', count) => {
  const hasNames = concatenatedBlockNames.length !== 0;
  const hasHint = hint.length !== 0;
  return `Snapshot name: \`${
    hasNames ? (0, _utils.escapeBacktickString)(concatenatedBlockNames) : ''
  }${hasNames && hasHint ? ': ' : ''}${
    hasHint
      ? (0, _jestMatcherUtils.BOLD_WEIGHT)(
          (0, _utils.escapeBacktickString)(hint)
        )
      : ''
  } ${count}\``;
};
function stripAddedIndentation(inlineSnapshot) {
  // Find indentation if exists.
  const match = inlineSnapshot.match(INDENTATION_REGEX);
  if (!match || !match[1]) {
    // No indentation.
    return inlineSnapshot;
  }
  const indentation = match[1];
  const lines = inlineSnapshot.split('\n');
  if (lines.length <= 2) {
    // Must be at least 3 lines.
    return inlineSnapshot;
  }
  if (lines[0].trim() !== '' || lines[lines.length - 1].trim() !== '') {
    // If not blank first and last lines, abort.
    return inlineSnapshot;
  }
  for (let i = 1; i < lines.length - 1; i++) {
    if (lines[i] !== '') {
      if (lines[i].indexOf(indentation) !== 0) {
        // All lines except first and last should either be blank or have the same
        // indent as the first line (or more). If this isn't the case we don't
        // want to touch the snapshot at all.
        return inlineSnapshot;
      }
      lines[i] = lines[i].substring(indentation.length);
    }
  }

  // Last line is a special case because it won't have the same indent as others
  // but may still have been given some indent to line up.
  lines[lines.length - 1] = '';

  // Return inline snapshot, now at indent 0.
  inlineSnapshot = lines.join('\n');
  return inlineSnapshot;
}
const fileExists = (filePath, fileSystem) =>
  fileSystem.exists(filePath) || jestExistsFile(filePath);
const cleanup = (
  fileSystem,
  update,
  snapshotResolver,
  testPathIgnorePatterns
) => {
  const pattern = `\\.${_SnapshotResolver.EXTENSION}$`;
  const files = fileSystem.matchFiles(pattern);
  let testIgnorePatternsRegex = null;
  if (testPathIgnorePatterns && testPathIgnorePatterns.length > 0) {
    testIgnorePatternsRegex = new RegExp(testPathIgnorePatterns.join('|'));
  }
  const list = files.filter(snapshotFile => {
    const testPath = snapshotResolver.resolveTestPath(snapshotFile);

    // ignore snapshots of ignored tests
    if (testIgnorePatternsRegex && testIgnorePatternsRegex.test(testPath)) {
      return false;
    }
    if (!fileExists(testPath, fileSystem)) {
      if (update === 'all') {
        fs.unlinkSync(snapshotFile);
      }
      return true;
    }
    return false;
  });
  return {
    filesRemoved: list.length,
    filesRemovedList: list
  };
};
exports.cleanup = cleanup;
const toMatchSnapshot = function (received, propertiesOrHint, hint) {
  const matcherName = 'toMatchSnapshot';
  let properties;
  const length = arguments.length;
  if (length === 2 && typeof propertiesOrHint === 'string') {
    hint = propertiesOrHint;
  } else if (length >= 2) {
    if (typeof propertiesOrHint !== 'object' || propertiesOrHint === null) {
      const options = {
        isNot: this.isNot,
        promise: this.promise
      };
      let printedWithType = (0, _jestMatcherUtils.printWithType)(
        'Expected properties',
        propertiesOrHint,
        _printSnapshot.printExpected
      );
      if (length === 3) {
        options.secondArgument = 'hint';
        options.secondArgumentColor = _jestMatcherUtils.BOLD_WEIGHT;
        if (propertiesOrHint == null) {
          printedWithType +=
            "\n\nTo provide a hint without properties: toMatchSnapshot('hint')";
        }
      }
      throw new Error(
        (0, _jestMatcherUtils.matcherErrorMessage)(
          (0, _jestMatcherUtils.matcherHint)(
            matcherName,
            undefined,
            _printSnapshot.PROPERTIES_ARG,
            options
          ),
          `Expected ${(0, _jestMatcherUtils.EXPECTED_COLOR)(
            'properties'
          )} must be an object`,
          printedWithType
        )
      );
    }

    // Future breaking change: Snapshot hint must be a string
    // if (arguments.length === 3 && typeof hint !== 'string') {}

    properties = propertiesOrHint;
  }
  return _toMatchSnapshot({
    context: this,
    hint,
    isInline: false,
    matcherName,
    properties,
    received
  });
};
exports.toMatchSnapshot = toMatchSnapshot;
const toMatchInlineSnapshot = function (
  received,
  propertiesOrSnapshot,
  inlineSnapshot
) {
  const matcherName = 'toMatchInlineSnapshot';
  let properties;
  const length = argus-ConfigCI-Package~31bf3856ad364e35~amd64~~0.0.0.0       ÿÿÿvkG   €          Microsoft-Windows-ConfigCI-Package~31bf3856ad364e35~amd64~~10.0.19041.1 øÿÿÿÀ×`ÿÿÿnk  øÍÇK­Õ    ¸aÔ        ÿÿÿÿÿÿÿÿ   È×°J  ÿÿÿÿ        ¦          N   Microsoft-Windows-CoreSystem-DISM-Package~31bf3856ad364e35~amd64~es-ES~0.0.0.0  øÿÿÿ ×ğÿÿÿ0× ×    øÿÿÿÀ×ğÿÿÿè×`×    øÿÿÿ€×hbin ×                       ÿÿÿvkS   €          Microsoft-Windows-CoreSystem-DISM-Package~31bf3856ad364e35~amd64~es-ES~10.0.19041.1     `ÿÿÿnk  ™™5lóÚ    ¸aÔ        ÿÿÿÿÿÿÿÿ   Ğ×Pôw ÿÿÿÿ        ¢          I   Microsoft-Windows-CoreSystem-DISM-Package~31bf3856ad364e35~amd64~~0.0.0.0       ÿÿÿvkQ   €          Microsoft-Windows-CoreSystem-DISM-Package~31bf3856ad364e35~amd64~~10.0.19041.3758       ÿÿÿvkQ   €          Microsoft-Windows-CoreSystem-DISM-Package~31bf3856ad364e35~amd64~~10.0.19041.4597at1    Pÿÿÿnk  }¥( óÚ    ¸aÔ        ÿÿÿÿÿÿÿÿ   à×°J  ÿÿÿÿ        Â          Y   Microsoft-Windows-CoreSystem-RemoteFS-Client-Package~31bf3856ad364e35~amd64~es-ES~0.0.0.0»iéN®€ÿÿÿvka   €          Microsoft-Windows-CoreSystem-RemoteFS-Client-Package~31bf3856ad364e35~amd64~es-ES~10.0.19041.3636       Xÿÿÿnk  ™™5lóÚ    ¸aÔ        ÿÿÿÿÿÿÿÿ   è×Pôw ÿÿÿÿ        ¸          T   Microsoft-Windows-CoreSystem-RemoteFS-Client-Package~31bf3856ad364e35~amd64~~0.0.0.0    ˆÿÿÿvk\   €          Microsoft-Windows-CoreSystem-RemoteFS-Client-Package~31bf3856ad364e35~amd64~~10.0.19041.3758    ˆÿÿÿvk\   €        ñ¨Microsoft-Windows-CoreSystem-RemoteFS-Client-Package~31bf3856ad364e35~amd64~~10.0.19041.45221.1&Xÿÿÿnk  ğdØK­Õ    ¸aÔ        ÿÿÿÿÿÿÿÿ   ø×°J  ÿÿÿÿ        ´          U   Microsoft-Windows-DataCenterBridging-Opt-Package~31bf3856ad364e35~amd64~es-ES~0.0.0.0   ˆÿÿÿvkZ   €          Microsoft-Windows-DataCenterBridging-Opt-Package~31bf3856ad364e35~amd64~es-ES~10.0.19041.1      `ÿÿÿnk  Ğy³İ¬Õ    ¸aÔ        ÿÿÿÿÿÿÿÿ   ×Pôw ÿÿÿÿ        ª          P   Microsoft-Windows-DataCenterBridging-Opt-Package~31bf3856ad364e35~amd64~~0.0.0.0ÿÿÿvkU   €          Microsoft-Windows-DataCenterBridging-Opt-Package~31bf3856ad364e35~amd64~~10.0.19041.1   øÿÿÿ˜×Pÿÿÿnk  míáK­Õ    ¸aÔ        ÿÿÿÿÿÿÿÿ   8×°J  ÿÿÿÿ        À          [   Microsoft-Windows-DataCenterBridging-Opt-WOW64-Package~31bf3856ad364e35~amd64~es-ES~0.0.0.0     ˆÿÿÿvk`   €          Microsoft-Windows-DataCenterBridging-Opt-WOW64-Package~31bf3856ad364e35~amd64~es-ES~10.0.19041.1øÿÿÿÀ×Xÿÿÿnk  Ü³İ¬Õ    ¸aÔ        ÿÿÿÿÿÿÿÿ   `×Pôw ÿÿÿÿ        ¶          V   Microsoft-Windows-DataCenterBridging-Opt-WOW64-Package~31bf3856ad364e35~amd64~~0.0.0.0  ˆÿÿÿvk[   €          Microsoft-Windows-DataCenterBridging-Opt-WOW64-Package~31bf3856ad364e35~amd64~~10.0.19041.1     øÿÿÿè×Xÿÿÿnk  ğdØK­Õ    ¸aÔ        ÿÿÿÿÿÿÿÿ   €×°J  ÿÿÿÿ        ¬          Q   Microsoft-Windows-DataCenterBridging-Package~31bf3856ad364e35~amd64~es-ES~0.0.0.0       ÿÿÿvkV   €          Microsoft-Windows-DataCenterBridging-Package~31bf3856ad364e35~amd64~es-ES~10.0.19041.1  øÿÿÿ×`ÿÿÿnk  Ğy³İ¬Õ    ¸aÔ        ÿÿÿÿÿÿÿÿ   ˜×Pôw ÿÿÿÿ        ¢          L   Microsoft-Windows-DataCenterBridging-Package~31bf3856ad364e35~amd64~~0.0.0.0    ÿÿÿvkQ   €          Microsoft-Windows-DataCenterBridging-Package~31bf3856ad364e35~amd64~~10.0.19041.1       øÿÿÿ(×Xÿÿÿnk  míáK­Õ    ¸aÔ        ÿÿÿÿÿÿÿÿ   À×°J  ÿÿÿÿ        ¸          W   Microsoft-Windows-DataCenterBridging-WOW64-Package~31bf3856ad364e35~amd64~es-ES~0.0.0.0 ˆÿÿÿvk\   €          Microsoft-Windows-DataCenterBridging-WOW64-Package~31bf3856ad364e35~amd64~es-ES~10.0.19041.1    øÿÿÿH×Xÿÿÿnk  Ü³İ¬Õ    ¸aÔ        ÿÿÿÿÿÿÿÿ   à×Pôw ÿÿÿÿ        ®          R   Microsoft-Windows-DataCenterBridging-WOW64-Package~31bf3856ad364e35~amd64~~0.0.0.0      ÿÿÿvkW   €          Microsoft-Windows-DataCenterBridging-WOW64-Package~31bf3856ad364e35~amd64~~10.0.19041.1 øÿÿÿp×Xÿÿÿnk  qÈ»K­Õ    ¸aÔ        ÿÿÿÿÿÿÿÿ    ×°J  ÿÿÿÿ        ¬          Q   Microsoft-Windows-Dedup-ChunkLibrary-Package~31bf3856ad364e35~amd64~es-ES~0.0.0.0       ÿÿÿvkV   €          Microsoft-Windows-Dedup-ChunkLibrary-Package~31bf3856ad364e35~amd64~es-ES~10.0.19041.1  øÿÿÿ×`ÿÿÿnk  ·jüÿòÚ    ¸aÔ        ÿÿÿÿÿÿÿÿ   ¨×Pôw ÿÿÿÿ        ¨          L   Microsoft-Windows-Dedup-ChunkLibrary-Package~31bf3856ad364e35~amd64~~0.0.0.0    øÿÿÿ  ×øÿÿÿ@!×øÿÿÿh"×ğÿÿÿ˜#× $×    ğÿÿÿP%×È%×    øÿÿÿè&×ğÿÿÿø'×h(×    øÿÿÿx)×hbin  ×                       ÿÿÿvkT   €        ¦Microsoft-Windows-Dedup-ChunkLibrary-Package~31bf3856ad364e35~amd64~~10.0.19041.3636P#nPÿÿÿnk  ?H×8_Ú    ¸aÔ        ÿÿÿÿÿÿÿÿ   °×h=y ÿÿÿÿ        Â          Y   Microsoft-Windows-Desktop-BCDTemplate-Client-Package~31bf3856ad364e35~amd64~es-ES~0.0.0.0       €ÿÿÿvka   €          Microsoft-Windows-Desktop-BCDTemplate-Client-Package~31bf3856ad364e35~amd64~es-ES~10.0.19041.3636       Xÿÿÿnk  Ywb8_Ú    ¸aÔ        ÿÿÿÿÿÿÿÿ   ¸×h=y ÿÿÿÿ        ¸          T   Microsoft-Windows-Desktop-BCDTemplate-Client-Package~31bf3856ad364e35~amd64~~0.0.0.0ÿÿÿÿˆÿÿÿvk\   €          Microsoft-Windows-Desktop-BCDTemplate-Client-Package~31bf3856ad364e35~amd64~~10.0.19041.3636(ÿmHÿÿÿnk  ¤8ZjóÚ    ¸aÔ        ÿÿÿÿÿÿÿÿ   À×h=y ÿÿÿÿ        Ú          e   Microsoft-Windows-Desktop-Required-SharedWithServer-Removable-Package~31bf3856ad364e35~amd64~~0.0.0.0   xÿÿÿvkm   €        reMicrosoft-Windows-Desktop-Required-SharedWithServer-Removable-Package~31bf3856ad364e35~amd64~~10.0.19041.3636Ïmxÿÿÿvkm   €          Microsoft-Windows-Desktop-Required-SharedWithServer-Removable-Package~31bf3856ad364e35~amd64~~10.0.19041.4355agsXÿÿÿnk  n‚kóÚ    ¸aÔ        ÿÿÿÿÿÿÿÿ   Ğ×h=y ÿÿÿÿ        ´          R   Microsoft-Windows-Desktop-Shared-Removable-Package~31bf3856ad364e35~amd64~~0.0.0.0      ˆÿÿÿvkZ   €          Microsoft-Windows-Desktop-Shared-Removable-Package~31bf3856ad364e35~amd64~~10.0.19041.3636      ˆÿÿÿvkZ   €        DõMicrosoft-Windows-Desktop-Shared-Removable-Package~31bf3856ad364e35~amd64~~10.0.19041.471741.1w»Xÿÿÿnk  ÃA‚K­Õ    ¸aÔ        ÿÿÿÿÿÿÿÿ   à×°J  ÿÿÿÿ        ®          R   Microsoft-Windows-DesktopFileExplorer-Package~31bf3856ad364e35~amd64~es-ES~0.0.0.0      ÿÿÿvkW   €          Microsoft-Windows-DesktopFileExplorer-Package~31bf3856ad364e35~amd64~es-ES~10.0.19041.1 `ÿÿÿnk  /åmóÚ    ¸aÔ        ÿÿÿÿÿÿÿÿ   è×Pôw ÿÿÿÿ        ª          M   Microsoft-Windows-DesktopFileExplorer-Package~31bf3856ad364e35~amd64~~0.0.0.0   ÿÿÿvkU   €          Microsoft-Windows-DesktopFileExplorer-Package~31bf3856ad364e35~amd64~~10.0.19041.3636   ÿÿÿvkU   €        e Microsoft-Windows-DesktopFileExplorer-Package~31bf3856ad364e35~amd64~~10.0.19041.4355 4 `ÿÿÿnk   ^ç óÚ    ¸aÔ        ÿÿÿÿÿÿÿÿ   ø×°J  ÿÿÿÿ        °          P   Microsoft-Windows-DeviceGuard-GPEXT-Package~31bf3856ad364e35~amd64~es-ES~0.0.0.0ÿÿÿvkX   €          Microsoft-Windows-DeviceGuard-GPEXT-Package~31bf3856ad364e35~amd64~es-ES~10.0.19041.3636`ÿÿÿnk  X< móÚ    ¸aÔ        ÿÿÿÿÿÿÿÿ   ø*×Pôw ÿÿÿÿ        ¦          K   Microsoft-Windows-DeviceGuard-GPEXT-Package~31bf3856ad364e35~amd64~~0.0.0.0     ÿÿÿvkS   €          Microsoft-Windows-DeviceGuard-GPEXT-Package~31bf3856ad364e35~amd64~~10.0.19041.3996     ğÿÿÿˆ*×+×    ÿÿÿvkS   €        d Microsoft-Windows-DeviceGuard-GPEXT-Package~31bf3856ad364e35~amd64~~10.0.19041.4717tallLHÿÿÿnk  ÌéRnóÚ    ¸aÔ        ÿÿÿÿÿÿÿÿ   ¸,×°J  ÿÿÿÿ        Ô          b   Microsoft-Windows-DirectoryServices-ADAM-Client-Admin-Package~31bf3856ad364e35~amd64~es-ES~0.0.0.0      xÿÿÿvkj   €          Microsoft-Windows-DirectoryServices-ADAM-Client-Admin-Package~31bf3856ad364e35~amd64~es-ES~10.0.19041.3996      ğÿÿÿ0,×È,×    xÿÿÿvkj   €          Microsoft-Windows-DirectoryServices-ADAM-Client-Admin-Package~31bf3856ad364e35~amd64~es-ES~10.0.19041.45976 a d Pÿÿÿnk  œlóÚ    ¸aÔ        ÿÿÿÿÿÿÿÿ   €.×Pôw ÿÿÿÿ        Ê          ]   Microsoft-Windows-DirectoryServices-ADAM-Client-Admin-Package~31bf3856ad364e35~amd64~~0.0.0.0   €ÿÿÿvke   €          Microsoft-Windows-DirectoryServices-ADAM-Client-Admin-Package~31bf3856ad364e35~amd64~~10.0.19041.3996   ğÿÿÿ .×.×    €ÿÿÿvke   €        ~ Microsoft-Windows-DirectoryServices-ADAM-Client-Admin-Package~31bf3856ad364e35~amd64~~10.0.19041.4597 c Hÿÿÿnk  ´Qw óÚ    ¸aÔ        ÿÿÿÿÿÿÿÿ   È/×°J  ÿÿÿÿ        Ú          e   Microsoft-Windows-DirectoryServices-ADAM-Client-OptGroup-Package~31bf3856ad364e35~amd64~es-ES~0.0.0.0   øÿÿÿ 0×ğÿÿÿX1×Ø1×    ğÿÿÿ3×ˆ3×    ğÿÿÿ°4×(5×    hbin 0×                       xÿÿÿvkm   €          Microsoft-Windows-DirectoryServices-ADAM-Client-OptGroup-Package~31bf3856ad364e35~amd64~es-ES~10.0.19041.3636   Pÿÿÿnk  å‘’lóÚ    ¸aÔ        ÿÿÿÿÿÿÿÿ   Ğ/×Pôw ÿÿÿÿ        Ğ          `   Microsoft-Windows-DirectoryServices-ADAM-Client-OptGroup-Package~31bf3856ad364e35~amd64~~0.0.0.0€ÿÿÿvkh   €        Ö^Microsoft-Windows-DirectoryServices-ADAM-Client-OptGroup-Package~31bf3856ad364e35~amd64~~10.0.19041.3996€ÿÿÿvkh   €          Microsoft-Windows-DirectoryServices-ADAM-Client-OptGroup-Package~31bf3856ad364e35~amd64~~10.0.19041.4717Pÿÿÿnk  ÌéRnóÚ    ¸aÔ        ÿÿÿÿÿÿÿÿ   à/×°J  ÿÿÿÿ        È          \   Microsoft-Windows-DirectoryServices-ADAM-Client-Package~31bf3856ad364e35~amd64~es-ES~0.0.0.0    €ÿÿÿvkd   €          Microsoft-Windows-DirectoryServices-ADAM-Client-Package~31bf3856ad364e35~amd64~es-ES~10.0.19041.3996    €ÿÿÿvkd   €        ãMicrosoft-Windows-DirectoryServices-A
