// eslint/lib/rules/no-mixed-operators.js
/**
 * @fileoverview Rule to disallow mixed binary operators.
 * @author Toru Nagashima
 * @deprecated in ESLint v8.53.0
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils.js");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

const ARITHMETIC_OPERATORS = ["+", "-", "*", "/", "%", "**"];
const BITWISE_OPERATORS = ["&", "|", "^", "~", "<<", ">>", ">>>"];
const COMPARISON_OPERATORS = ["==", "!=", "===", "!==", ">", ">=", "<", "<="];
const LOGICAL_OPERATORS = ["&&", "||"];
const RELATIONAL_OPERATORS = ["in", "instanceof"];
const TERNARY_OPERATOR = ["?:"];
const COALESCE_OPERATOR = ["??"];
const ALL_OPERATORS = [].concat(
    ARITHMETIC_OPERATORS,
    BITWISE_OPERATORS,
    COMPARISON_OPERATORS,
    LOGICAL_OPERATORS,
    RELATIONAL_OPERATORS,
    TERNARY_OPERATOR,
    COALESCE_OPERATOR
);
const DEFAULT_GROUPS = [
    ARITHMETIC_OPERATORS,
    BITWISE_OPERATORS,
    COMPARISON_OPERATORS,
    LOGICAL_OPERATORS,
    RELATIONAL_OPERATORS
];
const TARGET_NODE_TYPE = /^(?:Binary|Logical|Conditional)Expression$/u;

/**
 * Normalizes options.
 * @param {Object|undefined} options A options object to normalize.
 * @returns {Object} Normalized option object.
 */
function normalizeOptions(options = {}) {
    const hasGroups = options.groups && options.groups.length > 0;
    const groups = hasGroups ? options.groups : DEFAULT_GROUPS;
    const allowSamePrecedence = options.allowSamePrecedence !== false;

    return {
        groups,
        allowSamePrecedence
    };
}

/**
 * Checks whether any group which includes both given operator exists or not.
 * @param {Array<string[]>} groups A list of groups to check.
 * @param {string} left An operator.
 * @param {string} right Another operator.
 * @returns {boolean} `true` if such group existed.
 */
function includesBothInAGroup(groups, left, right) {
    return groups.some(group => group.includes(left) && group.includes(right));
}

/**
 * Checks whether the given node is a conditional expression and returns the test node else the left node.
 * @param {ASTNode} node A node which can be a BinaryExpression or a LogicalExpression node.
 * This parent node can be BinaryExpression, LogicalExpression
 *      , or a ConditionalExpression node
 * @returns {ASTNode} node the appropriate node(left or test).
 */
function getChildNode(node) {
    return node.type === "ConditionalExpression" ? node.test : node.left;
}

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

/** @type {import('../shared/types').Rule} */
module.exports = {
    meta: {
        deprecated: true,
        replacedBy: [],
        type: "suggestion",

        docs: {
            description: "Disallow mixed binary operators",
            recommended: false,
            url: "https://eslint.org/docs/latest/rules/no-mixed-operators"
        },

        schema: [
            {
                type: "object",
                properties: {
                    groups: {
                        type: "array",
                        items: {
                            type: "array",
                            items: { enum: ALL_OPERATORS },
                            minItems: 2,
                            uniqueItems: true
                        },
                        uniqueItems: true
                    },
                    allowSamePrecedence: {
                        type: "boolean",
                        default: true
                    }
                },
                additionalProperties: false
            }
        ],

        messages: {
            unexpectedMixedOperator: "Unexpected mix of '{{leftOperator}}' and '{{rightOperator}}'. Use parentheses to clarify the inten7.8C305.7 86.8 304.8 85.9 303.9 84.9L304.8 84C305.7 85 306.6 86 307.5 87ZM301.2 80.6L300.3 81.5C299.3 80.6 298.4 79.7 297.4 78.8L298.2 77.9C299.3 78.7 300.3 79.6 301.2 80.6ZM294.6 74.6L293.8 75.6C292.8 74.8 291.8 73.9 290.7 73.1L291.5 72.1C292.5 72.9 293.5 73.7 294.6 74.6ZM287.5 69.1L286.8 70.1C285.7 69.3 284.7 68.6 283.6 67.8L284.3 66.8C285.3 67.5 286.4 68.3 287.5 69.1ZM280.1 64.1L279.4 65.2C278.3 64.5 277.2 63.8 276 63.2L276.6 62.1C277.8 62.7 278.9 63.4 280.1 64.1ZM238.8 47.2L238.5 48.4C237.2 48.1 235.9 47.9 234.6 47.6L234.8 46.4C236.2 46.6 237.5 46.9 238.8 47.2ZM230 45.5L229.8 46.7C228.5 46.5 227.2 46.3 225.9 46.2L226 45C227.3 45.1 228.7 45.3 230 45.5ZM221.1 44.5L221 45.7C219.7 45.6 218.4 45.5 217.1 45.4L217.2 44.2C218.4 44.3 219.8 44.4 221.1 44.5ZM212.1 44V45.2C211.4 45.2 210.8 45.2 210.1 45.2C209.4 45.2 208.8 45.2 208.1 45.2V44C208.8 44 209.4 44 210.1 44C210.8 44 211.5 44 212.1 44ZM203.2 44.2L203.3 45.4C202 45.5 200.7 45.6 199.4 45.7L199.3 44.5C200.5 44.4 201.8 44.3 203.2 44.2ZM194.2 45L194.4 46.2C193.1 46.4 191.8 46.5 190.5 46.7L190.3 45.5C191.6 45.3 192.9 45.1 194.2 45ZM151.5 57.7L152.1 58.8C150.9 59.4 149.8 60 148.6 60.6L148 59.5C149.1 58.9 150.3 58.3 151.5 57.7ZM143.6 62L144.2 63.1C143.1 63.8 141.9 64.4 140.8 65.1L140.1 64C141.3 63.4 142.4 62.7 143.6 62ZM136 66.8C134.9 67.5 133.8 68.3 132.7 69.1L133.4 70.1C134.5 69.3 135.5 68.6 136.6 67.8L136 66.8ZM128.8 72.1C127.8 72.9 126.7 73.8 125.7 74.6L126.5 75.6C127.5 74.8 128.5 73.9 129.6 73.1L128.8 72.1ZM121.9 77.9L122.7 78.8C121.7 79.7 120.8 80.6 119.8 81.5L118.9 80.6C120 79.6 120.9 78.7 121.9 77.9ZM115.5 84.1L116.4 85C115.5 85.9 114.6 86.9 113.7 87.9L112.8 87.1C113.6 86 114.5 85 115.5 84.1ZM94.3 112.9L95.4 113.5C94.8 114.7 94.2 115.8 93.6 117L92.5 116.4C93.1 115.2 93.7 114.1 94.3 112.9ZM90.3 120.9L91.4 121.4C90.9 122.6 90.3 123.8 89.8 125L88.6 124.5C89.2 123.3 89.8 122.1 90.3 120.9ZM86.9 129.2L88.1 129.6C87.6 130.8 87.2 132.1 86.8 133.3L85.6 132.9C86 131.7 86.4 130.4 86.9 129.2ZM84 137.7L85.2 138.1C84.8 139.4 84.5 140.6 84.1 141.9L82.9 141.6C83.3 140.2 83.7 139 84 137.7ZM81.8 146.4L83 146.7C82.7 148 82.4 149.3 82.2 150.6L81 150.4C81.2 149 81.5 147.7 81.8 146.4ZM83 208.5L84.2 208.2C84.5 209.5 84.9 210.7 85.3 212L84.1 212.4C83.7 211.1 83.3 209.8 83 208.5ZM85.5 217.1C85.9 218.4 86.4 219.6 86.8 220.8L88 220.4C87.5 219.2 87.1 217.9 86.7 216.7L85.5 217.1ZM88.7 225.5L89.9 225C90.4 226.2 90.9 227.4 91.5 228.6L90.4 229.1C89.8 227.9 89.2 226.7 88.7 225.5ZM92.5 233.6L93.6 233C94.2 234.2 94.8 235.3 95.4 236.5L94.3 237.1C93.7 236 93.1 234.8 92.5 233.6ZM96.8 241.5L97.9 240.9C98.6 242 99.3 243.2 100 244.3L98.9 245C98.1 243.8 97.5 242.6 96.8 241.5ZM101.6 249L102.6 248.3C103.3 249.4 104.1 250.5 104.9 251.5L103.9 252.2C103.1 251.2 102.4 250.1 101.6 249ZM132.8 281L133.5 280C134.6 280.8 135.6 281.5 136.7 282.3L136 283.3C134.9 282.5 133.8 281.7 132.8 281ZM140.2 286C141.3 286.7 142.5 287.4 143.6 288.1L144.2 287C143.1 286.3 141.9 285.7 140.8 285L140.2 286ZM147.9 290.5C149.1 291.1 150.2 291.7 151.4 292.3L152 291.2C150.8 290.6 149.7 290 148.5 289.4L147.9 290.5ZM156 294.4L156.5 293.3C157.7 293.8 158.9 294.4 160.1 294.9L159.6 296.1C158.4 295.5 157.2 295 156 294.4ZM164.3 297.8L164.7 296.6C165.9 297.1 167.2 297.5 168.4 297.9L168 299.1C166.8 298.7 165.5 298.3 164.3 297.8ZM172.8 300.7L173.2 299.5C174.5 299.9 175.7 300.2 177 300.6L176.7 301.8C175.3 301.4 174 301 172.8 300.7ZM181.5 302.9L181.8 301.7C183.1 302 184.4 302.2 185.7 302.5L185.5 303.7C184.1 303.4 182.7 303.2 181.5 302.9ZM226 305.1L225.9 303.9C227.2 303.7 228.5 303.6 229.8 303.4L230 304.6C228.7 304.7 227.3 304.9 226 305.1ZM234.9 303.7C236.2 303.5 237.5 303.2 238.8 302.9L238.5 301.7C237.2 302 235.9 302.2 234.6 302.5L234.9
